### PR TITLE
Add std.encodeUTF8, std.decodeUTF8

### PR DESCRIPTION
**Pull request by sbarzowski**

### CODE REVIEW

File: core/vm.cpp [hunk 1]
```diff
@@ -1291,6 +1293,51 @@ class Interpreter {
...
+    const AST *builtinDecodeUTF8(const LocationRange &loc, const std::vector<Value> &args)
+    {
+        validateBuiltinArgs(loc, "decodeUTF8", args, {Value::ARRAY});
+
+        std::string byteString;
+
+        HeapArray *arr = static_cast<HeapArray *>(scratch.v.h);
+        for (auto *thunk : arr->elements) {
+            if (!thunk->filled) {
+                stack.newCall(loc, thunk, thunk->self, thunk->offset, thunk->upValues);
```
Comment (sbarzowski) [inline core/vm.cpp]: @sparkprime I need your help here. It doesn't work this way, and I'm not sure what is the right way. I need to get the (evaluated) elements of this array.

Comment (sparkprime) [reply core/vm.cpp]: I think you're going to need to record intermediate state on the stack while you recurse into the thunk elements. This is not easy.  Look at how std.filter is implemented natively -- it has a special stack frame it pushes.  If this puts you off implementing this natively in C++ I would understand it :)

Comment (sparkprime) [reply core/vm.cpp]: dmuir just did something similar for std.join in another PR

**Comment by sbarzowski** [comment]
@sparkprime I changed it to use frames and it seems to work. The @dmuir's work was indeed very helpful.

File: core/vm.cpp [hunk 2]
```diff
@@ -1308,6 +1314,61 @@ class Interpreter {
...
+    const AST *decodeUTF8(void)
+    {
+        Frame &f = stack.top();
+        const auto& elements = static_cast<HeapArray*>(f.val.v.h)->elements;
+        while (f.elementId < elements.size()) {
+            auto *th = elements[f.elementId];
+            if (th->filled) {
+                auto b = th->content;
+                if (b.t != Value::NUMBER) {
+                    throw makeError(stack.top().location, "Decoded byte was not a number");
```
[the remark this answers is not in the document]
Comment (sbarzowski) [reply core/vm.cpp]: Done

I've also added tests for these errors

**Pull request merged**